### PR TITLE
Added Support for Local Files

### PR DESCRIPTION
**Pull request by sleepychaser**
Parsing of the Spotify configuration for local files and the JSON returned by the web API.

## Changes
Local files imported by Spotify is stored in a custom binary format at `<Spotify` Config `Dir>/Users/<username>-user/local-files.bnk`. When playlist tracks are attempted to be loaded instead of discarding any local tracks attempt to find a local track in the `local-files.bnk` and if a matching title, artist, album, etc. are found then return a properly formatted Track structure

## TODO:
1. Test on platforms other than Linux
2. Attempting to play a local files throws unimplemented
3. Figure out how to decode and play formats other than Vorbis
4. Continue reverse engineering the local-files.bnk so that local files can be added through Psst
5. Maybe look into syncing local files with other devices on the network?

### CODE REVIEW

File: psst-gui/src/data/config.rs
Comment (jpochyla) [inline psst-gui/src/data/config.rs:169]: `map` would be better here I think?

File: psst-gui/src/main.rs
Comment (jpochyla) [inline psst-gui/src/main.rs:36]: This won't really work the first time user launches the app. Can we maybe inject the username (or the `LocalTrackManager`) into `WebApi` a bit later?

File: psst-gui/src/webapi/client.rs
Comment (jpochyla) [inline psst-gui/src/webapi/client.rs:55]: Again `map` I personally find preferable here, but maybe we should re-think how the manager is initialized, see above.

File: psst-gui/src/webapi/client.rs
Comment (jpochyla) [inline psst-gui/src/webapi/client.rs:381]: Is the endpoint really incorrect?

File: psst-gui/src/webapi/local.rs
Comment (jpochyla) [inline psst-gui/src/webapi/local.rs:70]: I really appreciate the docs!

File: psst-gui/src/webapi/local.rs
Comment (jpochyla) [inline psst-gui/src/webapi/local.rs:142]: Maybe using `String::from_utf8` directly, with `map_err`, would be better?

File: psst-gui/src/webapi/local.rs
Comment (jpochyla) [inline psst-gui/src/webapi/local.rs:262]: `?` should cover this I think?

File: psst-gui/src/data/track.rs
Comment (jpochyla) [inline psst-gui/src/data/track.rs:24]: `Arc<str>` would be preferable.

**Review by jpochyla — CHANGES_REQUESTED** [review]
Wow, nice! Added some nitpicks, I'd like to as well do some of my own modifications later, before merging this, if you don't mind.

**Comment by sleepychaser** [comment]
I committed most of the changes you suggested. The problem I was running into was injecting the username into the `WebApi` after log in as it becomes an Arc. I ended up just making the `LocalTrackManager` a singleton similar to `WebApi`. I don't know if this is the best way to do it and there may be a better way with Rust features I'm not familiar with. Feel free to make any changes you see fit.

**Comment by jpochyla** [comment]
Hi @sleepychaser, I've made some more changes (mostly some style nits, functionality should be identical I think). It works on macOS. The playback is next!

**Comment by jpochyla** [comment]
Re. multiple codecs support: this is definitely needed (not even all Spotify tracks are in Vorbis). Options:

1. Get rid of minivorbis, use [Symphonia](https://github.com/pdeljanov/Symphonia) (which seems *excellent*) for the decoding, and miniaudio just for the output.
2. Wait until miniaudio hits a stable version with the highlevel API and vorbis codec included.
3. Keep minivorbis and the MP3 decoder from miniaudio -- I don't like this much.

It's sad the the rust audio output is in such a bad state (rodio can not even keep the buffers full in debug mode, cpal has >100 issues open and maintainers don't seem to care, etc.), so it seems like we're stuck with a C library for this.

**Comment by jpochyla** [comment]
Merged to master, sorry about the delay :) Playback is still not supported, but now that we have Symphonia, it wouldn't be that hard to do.

**Comment by Kyiro** [comment]
Will playback ever be supported?

**Comment by jpochyla** [comment]
Hi @Kyiro, I've opened an issue for this. No ETA, though :(

**Pull request closed without merging**
